flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java

```diff
@@ -122,32 +122,32 @@ public RecordsWithSplitIds<ConsumerRecord<byte[], byte[]>> fetch() throws IOExce
         KafkaPartitionSplitRecords recordsBySplits =
                 new KafkaPartitionSplitRecords(consumerRecords, kafkaSourceReaderMetrics);
         List<TopicPartition> finishedPartitions = new ArrayList<>();
-        for (TopicPartition tp : consumerRecords.partitions()) {
+        for (TopicPartition tp : consumer.assignment()) {
             long stoppingOffset = getStoppingOffset(tp);
-            final List<ConsumerRecord<byte[], byte[]>> recordsFromPartition =
-                    consumerRecords.records(tp);
-
-            if (recordsFromPartition.size() > 0) {
-                final ConsumerRecord<byte[], byte[]> lastRecord =
-                        recordsFromPartition.get(recordsFromPartition.size() - 1);
-
-                // After processing a record with offset of "stoppingOffset - 1", the split reader
-                // should not continue fetching because the record with stoppingOffset may not
-                // exist. Keep polling will just block forever.
-                if (lastRecord.offset() >= stoppingOffset - 1) {
-                    recordsBySplits.setPartitionStoppingOffset(tp, stoppingOffset);
-                    finishSplitAtRecord(
-                            tp,
-                            stoppingOffset,
-                            lastRecord.offset(),
-                            finishedPartitions,
-                            recordsBySplits);
-                }
+            long consumerPosition = consumer.position(tp);
+            // Stop fetching when the consumer's position reaches the stoppingOffset.
+            // Control messages may follow the last record; therefore, using the last record's
+            // offset as a stopping condition could result in indefinite blocking.
+            if (consumerPosition >= stoppingOffset) {
```

Comment:
Afaik this is equivalent to the old `if (lastRecord.offset() >= stoppingOffset - 1) {`. This code gave the offset of the last method and your new version is using the offset of the next record.

Is this what you wanted? I'm not sure if this is fixing anything then except for the first fix above.

Reply:
As you mentioned, `consumerPosition >= stoppingOffset` is generally equivalent to `lastRecord.offset() >= stoppingOffset - 1`, but there are some edge cases where the new condition works better.

### Edge case
#### Producer setup
Let’s say we are producing transactional messages, and the `lastRecord's offset` is `100`. 
After producing the lastRecord, we commit the transaction, which writes a `marker message` at offset `101`. 
Now, the `nextRecord's offset` becomes `102` (one more than the last message including marker message).

#### Consumer setup
When the flink consumer is reading with `scan.bounded.mode` set to `latest-offset`, the `stoppingOffset` is set to the `nextRecord's offset`, which in this case is `102`.

#### Comparison
In the old condition `(lastRecord.offset() >= stoppingOffset - 1)`, it would compare `100 >= 102 - 1`, which would never be met, leading to indefinite blocking.
However, with the new condition `(consumerPosition >= stoppingOffset)`, it compares `102 >= 102`, which is immediately satisfied. 
This happens because consumer.position() skips the marker message. 
Using lastRecord.offset() in the stopping condition is insufficient in transactional case (when a marker message follows the last record).


| Condition       | As-Is (Old Condition)                             | To-Be (New Condition)                             |
|-----------------|---------------------------------------------------|---------------------------------------------------|
| Expression      | `lastRecord.offset() >= stoppingOffset - 1`        | `consumerPosition >= stoppingOffset`              |
| Example Comparison | Compares `100 >= 102 - 1`, which never meets     | Compares `102 >= 102`, which is immediately satisfied |
| Reason          | The condition fails because it checks the last data record's offset, not accounting for the marker message | The condition succeeds because `consumer.position()` skips the marker message and correctly evaluates the stopping condition |
| Issue           | Leads to indefinite polling                       | Stops polling immediately

Reply:
Thank you very much for the detailed explanation.